### DISCUSSION TITLE
Build Error

### DISCUSSION THREAD

**Discussion by stretch911** [PlummersSoftwareLLC/NightDriverStrip#214] [Q&A]
I updated my source code with latest release and now see squiggly lines under my #include statements.
when i build I get some python error?
What gives?
Processing atomlight (board: esp32dev; platform: espressif32; framework: arduino)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
'npm' is not recognized as an internal or external command,
operable program or batch file.
CalledProcessError: Command 'npm run build' returned non-zero exit status 1.:
  File "C:\Users\sever\.platformio\penv\Lib\site-packages\platformio\builder\main.py", line 181:
    env.SConscript(item, exports="env")
  File "C:\Users\sever\.platformio\packages\tool-scons\scons-local-4.4.0\SCons\Script\SConscript.py", line 597:
    return _SConscript(self.fs, *files, **subst_kw)
  File "C:\Users\sever\.platformio\packages\tool-scons\scons-local-4.4.0\SCons\Script\SConscript.py", line 285:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "C:\Users\sever\Documents\NightDriverStrip\bake_site.py", line 5:
    subprocess.check_output("npm run build", shell=True)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.496.0_x64__qbz5n2kfra8p0\Lib\subprocess.py", line 466:
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.496.0_x64__qbz5n2kfra8p0\Lib\subprocess.py", line 571:
    raise CalledProcessError(retcode, process.args,
============================================================================= [FAILED] Took 0.49 seconds =============================================================================

Environment    Status    Duration
-------------  --------  ------------
atomlight      FAILED    00:00:00.489
======================================================================== 1 failed, 0 succeeded in 00:00:00.48

**Comment by rbergen** [comment]
This is related to the work that's underway on the new on-device React-based web application/server. You'll need to make sure npm is installed on the build system so the web application can be built. At least on Windows it's part of Node.js, which can be downloaded from <https://nodejs.org/en/>.

And yes, it's true that this is not yet covered in the project README. I'll extend the README to mention this as soon as I have a chance.

**Reply by stretch911** [reply]
i've been keeping up w/ your work. looking forward to it!!!! thanks for info.

**Reply by stretch911** [reply]
i just installed node.js and rebooted vscode and still have the same issue... :(

**Reply by stretch911** [reply]
C:\Users\stretch>npm
npm <command>

Usage:

npm install        install all the dependencies in your project
npm install <foo>  add the <foo> dependency to your project
npm test           run this project's tests
npm run <foo>      run the script named <foo>
npm <command> -h   quick help on <command>
npm -l             display usage info for all commands
npm help <term>    search for help on <term> (in a browser)
npm help npm       more involved overview (in a browser)

All commands:

    access, adduser, audit, bugs, cache, ci, completion,
    config, dedupe, deprecate, diff, dist-tag, docs, doctor,
    edit, exec, explain, explore, find-dupes, fund, get, help,
    hook, init, install, install-ci-test, install-test, link,
    ll, login, logout, ls, org, outdated, owner, pack, ping,
    pkg, prefix, profile, prune, publish, query, rebuild, repo,
    restart, root, run-script, search, set, shrinkwrap, star,
    stars, start, stop, team, test, token, uninstall, unpublish,
    unstar, update, version, view, whoami

Specify configs in the ini-formatted file:
    C:\Users\stretch\.npmrc
or on the command line via: npm <command> --key=value

More configuration info: npm help config
Configuration fields: npm help 7 config

npm@9.4.0 C:\Program Files\nodejs\node_modules\npm

its installed....works on cmd line

**Reply by rbergen** [reply]
I looked into it a little further, and I know what the problem is. Sadly, the core problem is not that easy to fix, as it's related to the build process requiring a Unix-like environment, which Windows simply doesn't offer. In fact, I think we'll have to disable the automated web application build on Windows altogether. 
While I figure that out, the following should get you going with the current codebase, at least. In platformio.ini, you can put a semicolon in front of line 75, which should be `extra_scripts =  pre:bake_site.py`. It does mean the web application won't build, but it should allow you to compile the projects in themselves.

**Comment by stretch911** [comment]
i have a kali linux vm? should i work in that?

**Reply by rbergen** [reply]
I'm afraid I can't really comment on the extent to which Kali Linux, as I have no experience with it. I do run Windows as my OS, but I use WSL2 (running Ubuntu 22.04) for anything that prefers to run under Linux. I can confirm by personal experience that works well for building NightDriverStrip.

When reading your earlier message, I got the impression you may think I can take credit for the web application that's being developed. I just want to avoid any confusion: I can't in any way. The development on the web application is being done by @Louis-Riel. I'm just trying to manage things a little and help out where I can. :)

I do appreciate your patience with this. The new approach to the web application is still pretty new, and I think it's fair to say we're still in the process of ironing out the kinks.

**Reply by stretch911** [reply]
all good. wasn't specific. i just have seen you and Louis-Riel active in the code :)

**Reply by rbergen** [reply]
That's fine, and in that case: thank you! I just would hate to accept credit for something I didn't do.

I just committed a change to the bake_site.py script that makes the PlatformIO builds break on Windows. It now checks if it's running on Linux (WSL or otherwise) and bails out otherwise. That should fix the builds not working, at least.

**Reply by stretch911** [reply]
i don't see the option to install platformio ide extension in code-oss (vs code on kali)

**Reply by rbergen** [reply]
Considering that Kali is a distribution specifically put together for pen testing and other IT-sec tasks, I can imagine a fair amount of otherwise ubiquitous capabilities have been disabled in it.

**Reply by stretch911** [reply]
that fixed me up . muchos gracias

**Reply by stretch911** [reply]
it uploaded but i'm getting a bunch of puke in the serial monitor...stand by

**Reply by stretch911** [reply]
(I) (PrintOutputHeader)(C1) NightDriverStrip
(I)
(I) (PrintOutputHeader)(C1) ------------------------------------------------------------------------------------------------------------
(I) (PrintOutputHeader)(C1) M5STICKC: 0, USE_M5DISPLAY: 0, USE_OLED: 0, USE_TFTSPI: 0, USE_LCD: 0, USE_AUDIO: 1, ENABLE_REMOTE: 1
(I) (PrintOutputHeader)(C1) Version 36: Wifi SSID: eZone - ESP32 Free Memory: 234412, PSRAM:0, PSRAM Free: 0
(I) (PrintOutputHeader)(C1) ESP32 Clock Freq : 240 MHz
(I) (setup)(C1) Startup!
(I) (setup)(C1) Starting DebugLoopTaskEntry
(W) (StartDebugThread)(C1) >> Launching Debug Thread
(I) (DebugLoopTaskEntry)(C0) >> DebugLoopTaskEntry
(I)
(W) (setup)(C1) Reserving 145 LED buffers for a total of 107300 bytes...
(I) (setup)(C1) Adding LEDs to FastLED...
(I) (setup)(C1) Initializing effects manager...
(W) (InitEffectsManager)(C1) InitEffectsManager...
(I) (setup)(C1) Launching Drawing:
(E) (setup)(C1) Heap before launch: PASS
(W) (StartDrawThread)(C1) >> Launching Draw Thread
(W) (DrawLoopTaskEntad)(C1) >> Launching Remote Thread(W)
(W) (Setup)(C1) >> GFXBase::Setup
(W)
(W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
m(W) (RemoteLoopEntry)(C0) >> RemoteLoopEntry
(W) ((W) ed
read)(C1)C0)  >Remote Control Decoding hSrteaardt
W) ed
(I) ntry
cketThread)(C1) >> Launching Socket Thread
I) ntry
(I)
[   353][E][ESPmDNS.cpp:65] begin(): Failed starting MDNS
Error starting mDNS
(W) (StartAudioThread)(C1) >> Launching Audio Thread
(I) (AudioSamplerTaskEntry)(C1) >>> Sampler Task Started
(I) (setup)(C1) Setup complete - ESP32 Free Memory: 54008
(I)
(I) (ConnectToWiFi)(C0) Setting host name to NightRider...WL_NO_SHIELD
Pass 0 of 10: Connecting to Wifi SSID: eZone - ESP32 Free Memory: 54008, PSRAM:0, PSRAM Free: 0

abort() was called at PC 0x40085e59 on core 0


Backtrace: 0x400848c1:0x3fff91c0 0x4008e91d:0x3fff91e0 0x400942c9:0x3fff9200 0x40085e59:0x3fff9280 0x40085e86:0x3fff92a0 0x40085ffd:0x3fff92d0 0x40101fde:0x3fff92f0 0x4010126f:0x3fff9310 0x40100a6d:0x3fff9330 0x4008b0cd:0x3fff9350 0x40089dd2:0x3fff9370 0x40089e5c:0x3fff9390 0x4008a46d:0x3fff93c0 0x401952db:0x3fff93f0 0x4018ce4f:0x3fff9410 0x4018cfcd:0x3fff9720 0x401a6fad:0x3fff9750 0x401283d6:0x3fff9780 0x401284a2:0x3fff97b0 0x401a5eba:0x3fff9830 0x401a6021:0x3fff9860 0x401a64fa:0x3fff9890 0x401a651d:0x3fff98d0 0x401a5d29:0x3fff9900 0x401085fd:0x3fff9920 0x401086bf:0x3fff9940 0x40108733:0x3fff9960 0x400e7e6b:0x3fff9990 0x400e8029:0x3fff9b20 0x400e8268:0x3fff9b50 0x400e8f51:0x3fff9b70 0x400dbe4e:0x3fff9cd0 0x40081ec7:0x3fff9d20

  #0  0x400848c1:0x3fff91c0 in panic_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/panic.c:402
  #1  0x4008e91d:0x3fff91e0 in esp_system_abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_system/esp_system.c:128
  #2  0x400942c9:0x3fff9200 in abort at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/abort.c:46
  #3  0x40085e59:0x3fff9280 in lock_init_generic at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/locks.c:75
  #4  0x40085e86:0x3fff92a0 in lock_acquire_generic at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/locks.c:127
  #5  0x40085ffd:0x3fff92d0 in _lock_acquire_recursive at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/newlib/locks.c:167
  #6  0x40101fde:0x3fff92f0 in uart_write at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/vfs/vfs_uart.c:200
  #7  0x4010126f:0x3fff9310 in console_write at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/vfs/vfs_console.c:73
  #8  0x40100a6d:0x3fff9330 in esp_vfs_write at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/vfs/vfs.c:431 (discriminator 4)
  #9  0x4008b0cd:0x3fff9350 in __swrite at /builds/idf/crosstool-NG/.build/HOST-x86_64-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/stdio.c:94
  #10 0x40089dd2:0x3fff9370 in __sflush_r at /builds/idf/crosstool-NG/.build/HOST-x86_64-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/fflush.c:224
  #11 0x40089e5c:0x3fff9390 in _fflush_r at /builds/idf/crosstool-NG/.build/HOST-x86_64-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/fflush.c:278
  #12 0x4008a46d:0x3fff93c0 in __sfvwrite_r at /builds/idf/crosstool-NG/.build/HOST-x86_64-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/fvwrite.c:251
  #13 0x401952db:0x3fff93f0 in __sprint_r at /builds/idf/crosstool-NG/.build/HOST-x86_64-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/vfprintf.c:433
      (inlined by) __sprint_r at /builds/idf/crosstool-NG/.build/HOST-x86_64-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/vfprintf.c:403
  #14 0x4018ce4f:0x3fff9410 in _vfprintf_r at /builds/idf/crosstool-NG/.build/HOST-x86_64-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/vfprintf.c:1781 (discriminator 1)  
  #15 0x4018cfcd:0x3fff9720 in vprintf at /builds/idf/crosstool-NG/.build/HOST-x86_64-w64-mingw32/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/vprintf.c:34 (discriminator 5)
  #16 0x401a6fad:0x3fff9750 in esp_log_writev at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/log/log.c:200
  #17 0x401283d6:0x3fff9780 in __wrap_esp_log_writev at /Users/ficeto/Desktop/ESP32/ESP32S2/esp32-arduino-lib-builder/components/esp-insights/components/esp_diagnostics/src/esp_diagnostics_log_hook.c:460
  #18 0x401284a2:0x3fff97b0 in __wrap_esp_log_write at /Users/ficeto/Desktop/ESP32/ESP32S2/esp32-arduino-lib-builder/components/esp-insights/components/esp_diagnostics/src/esp_diagnostics_log_hook.c:477
  #19 0x401a5eba:0x3fff9830 in base_node_add_handler at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_event/esp_event.c:235 (discriminator 3)
  #20 0x401a6021:0x3fff9860 in loop_node_add_handler at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_event/esp_event.c:317
  #21 0x401a64fa:0x3fff9890 in esp_event_handler_register_with_internal at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_event/esp_event.c:750
  #22 0x401a651d:0x3fff98d0 in esp_event_handler_register_with at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_event/esp_event.c:761
  #23 0x401a5d29:0x3fff9900 in esp_event_handler_register at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_event/default_event_loop.c:32
  #24 0x401085fd:0x3fff9920 in _esp_wifi_set_default_wifi_handlers at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_wifi/src/wifi_default.c:163
      (inlined by) _esp_wifi_set_default_wifi_handlers at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_wifi/src/wifi_default.c:156
  #25 0x401086bf:0x3fff9940 in esp_wifi_set_default_wifi_sta_handlers at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_wifi/src/wifi_default.c:217
  #26 0x40108733:0x3fff9960 in esp_netif_create_default_wifi_ap at /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/esp_wifi/src/wifi_default.c:321
  #27 0x400e7e6b:0x3fff9990 in wifiLowLevelInit(bool) at C:/Users/sever/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFiGeneric.cpp:661
  #28 0x400e8029:0x3fff9b20 in WiFiGenericClass::mode(wifi_mode_t) at C:/Users/sever/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFiGeneric.cpp:1149
  #29 0x400e8268:0x3fff9b50 in WiFiGenericClass::enableSTA(bool) at C:/Users/sever/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFiGeneric.cpp:1231
  #30 0x400e8f51:0x3fff9b70 in WiFiSTAClass::begin(char const*, char const*, int, unsigned char const*, bool) at C:/Users/sever/.platformio/packages/framework-arduinoespressif32/libraries/WiFi/src/WiFiSTA.cpp:226
  #31 0x400dbe4e:0x3fff9cd0 in ConnectToWiFi(unsigned int) at src/network.cpp:250
  #32 0x40081ec7:0x3fff9d20 in NetworkHandlingLoopEntry(void*) at src/main.cpp:309 (discriminator 1)




ELF file SHA256: aaf45460171785cb

E (2411) esp_core_dump_flash: Core dump flash config is corrupted! CRC=0x7bd5c66f instead of 0x0
Rebooting...
ets Jul 29 2019 12:21:46

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13104
load:0x40080400,len:3036
entry 0x400805e4
E (520) I2S: i2s_calculate_adc_dac_clock(1046): sample rate is too large
E (520) I2S: i2s_calculate_clock(1199): ADC clock calculate failed
E (524) esp_core_dump_flash: N���ɕ�dump partition found!
E (527) esp_core_dump_flash: No core dump partition found!
Replacing Idle Tasks with TaskManager...
(I) (PrintOutputHeader)(C1) NightDriverStrip
(I)
(I) (PrintOutputHeader)(C1) ------------------------------------------------------------------------------------------------------------
(I) (PrintOutputHeader)(C1) M5STICKC: 0, USE_M5DISPLAY: 0, USE_OLED: 0, USE_TFTSPI: 0, USE_LCD: 0, USE_AUDIO: 1, ENABLE_REMOTE: 1
(I) (PrintOutputHeader)(C1) Version 36: Wifi SSID: eZone - ESP32 Free Memory: 234412, PSRAM:0, PSRAM Free: 0
(I) (PrintOutputHeader)(C1) ESP32 Clock Freq : 240 MHz
(I) (setup)(C1) Startup!
(I) (setup)(C1) Starting DebugLoopTaskEntry
(W) (StartDebugThread)(C1) >> Launching Debug Thread
(I) (DebugLoopTaskEntry)(C0) >> DebugLoopTaskEntry
(I)
(W) (setup)(C1) Reserving 145 LED buffers for a total of 107300 bytes...
(I) (setup)(C1) Adding LEDs to FastLED...
(I) (setup)(C1) Initializing effects manager...
(W) (InitEffectsManager)(C1) InitEffectsManager...
(I) (setup)(C1) Launching Drawing:
(E) (setup)(C1) Heap before launch: PASS
(W) (StartDrawThread)(C1) >> Launching Draw Thread
(W) (DrawLoopTaskEntad)(C1) >> Launching Remote Thread(W)
(W) (Setup)(C1) >> GFXBase::Setup
(W)
(W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
m(W) (RemoteLoopEntry)(C0) >> RemoteLoopEntry
(W)
(W) ted
a(db)e(gCi1n)0) Remote Control Decodgh rSetaadr
W) ted
(I) ntry
cketThread)(C1) >> Launching Socket Thread
I) ntry
(I)
[   353][E][ESPmDNS.cpp:65] begin(): Failed starting MDNS
Error starting mDNS
(W) (StartAudioThread)(C1) >> Launching Audio Thread
(I) (AudioSamplerTaskEntry)(C1) >>> Sampler Task Started
(I) (setup)(C1) Setup complete - ESP32 Free Memory: 54008
(I)

**Reply by rbergen** [reply]
@Louis-Riel I'm feel I'm running out of steam when it comes to analyzing this, other than seeing hints that this is related to the device trying to run a web application that doesn't exist. Could you lend your view on what's happening here?

**Reply by rbergen** [reply]
Louis and I had an out-of-band exchange about this, and he's going to change the website build process such that it will only require Python (which is included with PlatformIO), and no longer depend on npm or POSIX tools. 

In the mean time, on Windows, at least for projects that have the webserver enabled, the website needs to be built manually, before the overall image is built. The instructions for that are in the file [site/README.md](https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/site/README.md).
